numba/stencil.py

```diff
@@ -0,0 +1,617 @@
+                        rvar = ir.Var(scope, "out", loc)
+                        # Write the return statements original value into
+                        # the array using the tuple index.
+                        si = ir.SetItem(rvar, s_index_var, stmt.value, loc)
+                        new_body.append(si)
+                else:
+                    new_body.append(stmt)
+            block.body = new_body
+
+    def add_indices_to_kernel(self, kernel, ndim, neighborhood, standard_indexed):
+        """
+        Transforms the stencil kernel as specified by the user into one
+        that includes each dimension's index variable as part of the getitem
+        calls.  So, in effect array[-1] becomes array[index0-1].
+        """
+        const_dict = {}
+        kernel_consts = []
+
+        if config.DEBUG_ARRAY_OPT == 1:
+            print("add_indices_to_kernel", ndim, neighborhood)
+            ir_utils.dump_blocks(kernel.blocks)
+
+        need_to_calc_kernel = False
+        if neighborhood is None:
+            need_to_calc_kernel = True
+        else:
+            if len(neighborhood) != ndim:
+                raise ValueError("%d dimensional neighborhood specified for %d dimensional input array" % (len(self.neighborhood), ndim))
+
+        tuple_table = ir_utils.get_tuple_table(kernel.blocks)
+
+        relatively_indexed = set()
+
+        for block in kernel.blocks.values():
+            scope = block.scope
+            loc = block.loc
+            new_body = []
+            for stmt in block.body:
+                if isinstance(stmt, ir.Assign) and isinstance(stmt.value, ir.Const):
+                    if config.DEBUG_ARRAY_OPT == 1:
+                        print("remembering in const_dict", stmt.target.name, stmt.value.value)
+                    # Remember consts for use later.
+                    const_dict[stmt.target.name] = stmt.value.value
+                if ((isinstance(stmt, ir.Assign)
+                        and isinstance(stmt.value, ir.Expr)
+                        and stmt.value.op in ['setitem', 'static_setitem']
+                        and stmt.value.value.name in kernel.arg_names) or 
+                   (isinstance(stmt, ir.SetItem)
+                        and stmt.target.name in kernel.arg_names)):
+                    raise ValueError("Assignments to arrays passed to stencil kernels is not allowed.")
+                if (isinstance(stmt, ir.Assign)
+                        and isinstance(stmt.value, ir.Expr)
+                        and stmt.value.op in ['getitem', 'static_getitem']
+                        and stmt.value.value.name in kernel.arg_names
+                        and stmt.value.value.name not in standard_indexed):
+                    # We found a getitem from the input array.
+                    if stmt.value.op == 'getitem':
+                        stmt_index_var = stmt.value.index
+                    else:
+                        stmt_index_var = stmt.value.index_var
```

Comment:
This line is unreachable and it should raise an exception instead.  This branch corresponds to when a `static_getitem` is seen but that does not happen because:

* `RewriteConstGetitems` is not used to get the kernel IR;
* only tuple unpack bytecode would emit this by default.

Also, demoting a `static_getitem` to a regular `getitem` seems dangerous.  It safer to error out on an unexpected `static_getitem`.

Reply:
Fixed.